Beta test registration

omarkhan opened the pull request:
This pull request adds a registration form for the OpenCraft beta. The form is available at `/beta/register/`.

#### Testing

- Checkout this branch
- Run the migrations
- In a separate terminal window inside the development VM, start a dummy smtp server with `sudo python -m smtpd -n -c DebuggingServer localhost:25`
- Start the development server
- Navigate to http://localhost:5000/beta/register/
- Play around with the input fields. Error messages should appear for invalid input
- Submit the form. You should see 2 verification emails in the smtp server output
- Follow the links in the verification emails
- In the django admin, check that a registration, user, and profile have been created, and that the email addresses have been verified

Note that I have not done any styling. This will be done by a web design agency next month.

omarkhan commented:
@smarnach this is ready for a look now.

omarkhan commented:
Oops, @itsjeyd this one is for you

smarnach commented on betatest/forms.py:
```diff
@@ -0,0 +1,205 @@
+        fields = self.fields.copy()
+        self.fields = OrderedDict()
+        for field in self.field_order:
+            self.fields[field] = fields[field]
```
Why is this necessary?  I think fields are rendered in the order in which you define them in the form class, so you could just define them in the correct order.  (I vaguely remember they keep a global counter when a Field class is instantiated, and use that global counter for ordering.)
In any case, you don't need to copy `self.fields`, since you don't even modify it, and even if you did that would be OK, since the fields shared between all instances of the class are in `self.base_fields`.

smarnach replied:
I see, some of the fields come from the model, and some of the fields are defined here, that makes sense.

(I just had a quick look at the code because I'm curious, by the way.  I'm not actually doing a review.)

antoviaque commented on betatest/templates/betatest/application.html:
```diff
@@ -0,0 +1,38 @@
+            <h1>Register for Beta</h1>
+            <p>This is the application form for the beta of the self-service
+            hosting offered by OpenCraft. For enterprise or institutional
+            offerings, contact us.</p>
```
Can you make "contact us" a link to contact@opencraft.com ?

(Just lurking quickly too : ) The PR looks neat! Glad to see the integration of forms with angular : ) )

smarnach commented on betatest/models.py:
```diff
@@ -0,0 +1,102 @@
+        validators=[
+            validators.RegexValidator(
+                r'^[\w.]+$',
+                'Please include only letters, numbers, and the . character.',
```
The regex also allows `_`.  Shouldn't we allow hyphens as well?  they are quite common in domain names.

smarnach commented on betatest/forms.py:
```diff
@@ -0,0 +1,205 @@
+    _subdomain_field = Meta.model._meta.get_field('subdomain')
+    _subdomain_validator = next(v for v in _subdomain_field.validators
+                                if hasattr(v, 'regex'))
+    subdomain = forms.RegexField(
```
Do we validate whether the subdomain is already taken?  We should look at both the instances that are currently running and the other beta test applications.  It might also make sense to blacklist certain values (we don't really want anyone to register `www.openedxhosting.com`), but we could do that manually for now.

omarkhan replied:
@smarnach we do validate whether the subdomain is already taken, but only by other beta registrations, not by instances that are currently running. I will add a validation step for that.

omarkhan commented:
I'm having some trouble writing js tests. The way they are currently set up, they are run completely in isolation, without any of the html from the django templates. This doesn't really make sense here, as most of the functionality I want to test relates to the form, and what django-angular does with it. @itsjeyd @smarnach @antoviaque what should I do here? Just skip these and focus on selenium tests? Extend the js tests to include a fake DOM with something like [jsdom](https://github.com/tmpvar/jsdom)?

itsjeyd commented on betatest/admin.py:
```diff
@@ -0,0 +1,36 @@
+# -*- coding: utf-8 -*-
+#
+# OpenCraft -- tools to aid developing and hosting free software projects
+# Copyright (C) 2015 OpenCraft <xavier@opencraft.com>
```
@omarkhan This should probably be updated to say `2015-2016` (also for the other files you added).

smarnach replied:
Do we really need copyright notes in every file?  Google stopped doing this since their lawyers said that they don't have any legal relevance whatsoever, and that having a single license file in the project achieves exactly the same.  (I can't judge whether that's true, but I find having these redundant comments everywhere rather annoying.)

bradenmacdonald replied:
I know Xavier likes these but I feel the same way :)

antoviaque replied:
I don't know either, but in doubt on licensing issues, I would pick the opinion of the FSF over the one from Google. Here in any case it seem minor, and it doesn't have much inconvenience to it. I don't personally see that section negatively - it's a place we talk to other devs or companies looking to work on it, and can even be part of the visual esthetic : )

However we should change the contact email - to contact@opencraft.com : )

smarnach replied:
All the section on how to apply the licence says is

> It is safest to attach them to the start of each source file to most effectively state the exclusion of warranty.

It doesn't say this is necessary for the licence to be valid or anything.

(And I wouldn't trust legal advice from the FSF too much.  It seems to be generally agreed that substantial parts of the AGPL are unenforcable under US law; at least that's what I heard from an FSF member.)

antoviaque replied:
@smarnach Are you sure trolling about the enforceability of the GPL over the format of headers is a good use of our time here? I'd be happy to chat about the FSF and the GPL around a beer though, and I will definitely be interested in knowing which parts of the GPL are unenforceable - once you sprint is clean can you send me the details?

Also, besides the technical legal skills, there are other elements to consider when considering legal advice, such as the motivations of the organization giving them, and how this might fit their larger strategy. I don't always agree with everything the FSF does or say, but at least their motivations are clear and strongly held to. They align in many ways with ours. On the other side, on licensing Google is quite opposed to many of the principles of free software - cf the fact that the vast majority of the code they produce is proprietary, and their position on software patents (which is behind a lot of their opposition to the GPLv3, btw).

smarnach replied:
@antoviaque My main intention in the last comment was to point out that even the FSF recommendation to include the header everywhere isn't that strongly worded.  I didn't mean to imply any critique of the goals of the FSF in general, which I definitely agree with.

itsjeyd replied:
@antoviaque @smarnach @omarkhan If we are keeping the notices, should the years be updated?

omarkhan replied:
Whatever we decide here, can we do it in a separate PR?

itsjeyd replied:
@omarkhan I'd be fine with that, since we'll also have to update the notices in existing files. Let's see what @antoviaque and @smarnach think.

smarnach replied:
My suggestion would be to change the email address and remove the year in a separate PR.  The year is bound to get outdated anyway.

antoviaque replied:
Yup, updating the email (to contact@opencraft.com) in a separate PR works for me.

antoviaque replied:
Done in https://github.com/open-craft/opencraft/pull/6/commits/cb0cf5df710789a5eedbb12eca33bfd1a8d54a38 . Updating the year only takes running a sed command, so I've left it, and changed it to `2015-2016` in all files.

itsjeyd commented on betatest/forms.py:
```diff
@@ -0,0 +1,205 @@
+class BetaTestApplicationForm(NgModelFormMixin, NgFormValidationMixin, NgModelForm):
+    """
+    Application form for beta testers. Creates User, UserProfile, and
+    BetaTestApplication models.
```
@omarkhan Nit: "Creates instances of User, UserProfile, and BetaTestApplication models on submit."

itsjeyd commented on betatest/models.py:
```diff
@@ -0,0 +1,102 @@
+        verbose_name='',
+        help_text=('I want OpenCraft to keep me updated about the progress '
+                   'of the beta test, and send me an email occasionally '
+                   'about it.'),
```
@omarkhan Nit: "and occasionally send me an email about it" maybe?

itsjeyd commented on betatest/models.py:
```diff
@@ -0,0 +1,102 @@
+                   '\n\nThis needs to be a valid email.'),
+    )
+    project_description = models.TextField(
+        verbose_name='your project',
```
@omarkhan Capitalize first letter? "Your project"

omarkhan replied:
Django does this automatically when displaying a form. The [example in the docs](https://docs.djangoproject.com/en/1.8/ref/models/options/#verbose-name) is lowercase, so I followed that convention.

itsjeyd commented on betatest/models.py:
```diff
@@ -0,0 +1,102 @@
+    )
+    subscribe_to_updates = models.BooleanField(
+        default=False,
+        verbose_name='',
```
@omarkhan Any specific reason for setting this to `''`?

omarkhan replied:
There is no label on the mockup. I tried setting this to `None`, but that just causes django to fall back to its default behaviour of using the field name as a label. The empty string seems to work. I will add a comment to clarify

itsjeyd commented on betatest/tests.py:
```diff
@@ -0,0 +1,158 @@
+            'public_contact_email': 'support@hogwarts.edu',
+            'password': 'gryffindor',
+            'password_confirmation': 'gryffindor',
+            'project_description': 'Online courses in Witchcraft and Wizardry',
```
@omarkhan Where can I sign up? :D

itsjeyd commented on betatest/tests.py:
```diff
@@ -0,0 +1,158 @@
+        self.assertEqual(profile.full_name, self.application['full_name'])
+
+        # Test email verification flow
+        self.assertEqual(len(mail.outbox), 2)
```
@omarkhan Should we check if email is *unverified* before clicking the links below?

itsjeyd commented on email_verification/models.py:
```diff
@@ -0,0 +1,145 @@
+        return instance
+
+    @classmethod
+    def check_verification_code(cls, code):
```
@omarkhan I think it would make sense to rename this method to `verify_mail`, to better reflect its main purpose. The fact that it checks the verification code is sort of incidental (it might be doing other checks as well before marking an email address as valid).

omarkhan replied:
I'm ok to rename this, but then we should also rename the `verify` method above. What would you call that?

itsjeyd replied:
@omarkhan Good point, how about `send_verification_email`?

omarkhan replied:
That works. Done

itsjeyd commented on email_verification/models.py:
```diff
@@ -0,0 +1,145 @@
+        address as verified if it is. Returns the corresponding
+        `EmailVerification` instance.
+        """
+        instance = cls.objects.get(verification_code=code)
```
@omarkhan `instance` is a bit generic (and slightly confusing since we also have an `instance` model). Rename to `email_verification`, perhaps?

itsjeyd commented on email_verification/templates/email_verification/verify.html:
```diff
@@ -0,0 +1,16 @@
+    {% if verification.verified %}
+    Thank you for verifying your email address ({{ verification.email }}).
+    {% elif verification.verification_code_expired %}
+    {# TODO: Resend verification email #}
```
@omarkhan Leftover TODO, just pointing out to make sure we don't forget about it.

itsjeyd commented on email_verification/templates/email_verification/verify.html:
```diff
@@ -0,0 +1,16 @@
+    {# TODO: Resend verification email #}
+    Your email verification link has expired.
+    {% else %}
+    Your email address could not be verified - please contact us.
```
@omarkhan Make "contact us" a link here to when you address @antoviaque's comment :)

antoviaque replied:
+1 :)

itsjeyd commented on email_verification/tests/test_models.py:
```diff
@@ -0,0 +1,117 @@
+        email = mail.outbox[-1]
+        self.assertIn(verification.verification_code, email.body)
+
+    def test_check_verification_code(self):
```
@omarkhan This will need to be updated as well when you rename `check_verification_code` to `verify_email`.

itsjeyd commented on email_verification/tests/test_views.py:
```diff
@@ -0,0 +1,73 @@
+
+    def test_verify_email(self):
+        """
+        Test that we can check a verification code.
```
@omarkhan "... that we can verify an email address."

itsjeyd commented on email_verification/tests/test_views.py:
```diff
@@ -0,0 +1,73 @@
+    @patch.object(EmailVerification, 'verification_code_expiry', timedelta(0))
+    def test_verify_email_code_expired(self):
+        """
+        Test attempting to verify an expired verification code.
```
@omarkhan Nit: "... to verify an email address with an expired verification code."

antoviaque commented on betatest/urls.py:
```diff
@@ -0,0 +1,37 @@
+urlpatterns = [
+    url(r'^register/$', views.BetaTestApplicationView.as_view(), name='register'),
+    url(r'^success/$', views.success, name='success'),
+    url(r'^api/register/validate/$', views.validate_registration, name='validate'),
```
Would you be able to use the standard API URL pattern & listing registration here? Cf https://github.com/open-craft/opencraft/tree/master/api

omarkhan replied:
@antoviaque you want me to move this into the api app? I don't think that makes sense, as the functionality is completely specific to this app, used only for ajax validation of the registration form. If we move it into the api app we create an unnecessary coupling between apps.

antoviaque replied:
@omarkhan If you look at the link I gave, you will see that this module doesn't contain code (which is always in the app itself, cf how it's done for the instance app), it's just a place where we register the api router - the idea being to have a registry that can list all the API calls supported, as well as a standard URL pattern: http://console.opencraft.com/api/v1/

omarkhan replied:
I understand, but this is an internal api that is used for something very specific. If you think we should expose this endpoint anyway, I can add it to the `api` urlconf, but I would prefer to keep it here as well to keep the `betatest` app self-contained.

antoviaque replied:
@omarkhan I'd like to follow the same pattern for all APIs - they are currently all internal, only meant to be used by the browser client.

omarkhan replied:
@antoviaque done, see https://github.com/open-craft/opencraft/pull/59/commits/f2a2931ac6232e574a36767d886c91de04a336b1. It feels like a bit of a hack but it works.

antoviaque replied:
@omarkhan Thanks! Does it make the API call discoverable through the API web interface? 

Also, why does it feel like a hack to you? It does introduce a dependency between the api app and the beta app, but the api app is used solely for managing the API URL scheme, so that seem fine. Or is there a better way to manage a common API scheme, including disoverability & browsable web interface, which I missed?

omarkhan replied:
Yes this makes it discoverable through the browsable api web interface.

I'm not happy with the coupling between apps, as it means that we can no longer move the app into its own django project, for example. But that is not why this feels like a hack to me.

It feels like a hack because DRF routers only support viewsets, which are an abstraction designed to automatically generate a RESTful api for models. In this case we are simply validating a form, models are not involved. So to get this to work with the browsable api I had to repurpose the viewset's `list` method, even though we are not listing any models. That feels like a hack to me because we are not using the abstraction the way it was designed. It just happens to work.

The quick and easy way I have done this in the past is to simply list the available endpoints in the project's documentation. Then you can visit the url in your browser and see DRF's autogenerated docs.

A more complete solution would be to use something like [DRF Docs](https://github.com/ekonstantinidis/django-rest-framework-docs). Other options are listed here: http://www.django-rest-framework.org/topics/documenting-your-api/

antoviaque replied:
@omarkhan Thanks for the details. DRF Docs looks neat, it's a good idea imho - do you want to add it to the uservoice, so we consider it to replace the current approach in the future?

omarkhan replied:
Sounds good, I will do that

antoviaque commented on email_verification/models.py:
```diff
@@ -0,0 +1,145 @@
+
+# Models ######################################################################
+
+class EmailVerification(models.Model):
```
Was there an advantage to reimplement this over reusing an existing one, like https://github.com/mfogel/django-simple-email-confirmation ?

omarkhan replied:
@antoviaque I evaluated [django-simple-email-confirmation](https://github.com/mfogel/django-simple-email-confirmation), [django-registration](https://github.com/macropin/django-registration), and [django-userena](https://github.com/bread-and-pepper/django-userena) for this. All of them implement email verification but tie email addresses to the `User` model. In this case we also want to verify an email address that belongs to the `BetaTestApplication` model. I considered using django-simple-email-confirmation and adding some mappings between its `EmailAddress` model and out `BetaTestApplication` model, but in the end I decided it would be cleaner and easier to just implement the basic email verification functionality as a separate app.

antoviaque replied:
@omarkhan What about contributing the ability to confirm multiple addresses to one of these upstream projects? Email verification is a very standard feature, the ecosystem would likely benefit more from improving the projects that already exist rather than adding our own.

omarkhan replied:
@antoviaque OK I will work on a PR for django-simple-email-confirmation once I have finished the selenium tests for the registration form.

smarnach commented:
@omarkhan Sorry, I can't help regarding the question how to implement JS tests for this, that's a topic I should learn more about.

antoviaque commented:
@omarkhan 

>I'm having some trouble writing js tests. The way they are currently set up, they are run completely in isolation, without any of the html from the django templates. This doesn't really make sense here, as most of the functionality I want to test relates to the form, and what django-angular does with it. @itsjeyd @smarnach @antoviaque what should I do here? Just skip these and focus on selenium tests? Extend the js tests to include a fake DOM with something like jsdom?

I'm not completely sure I understand what is blocking you - you have the dom of the browser? But yes you need to test the result of the operations based on changes on a DOM fixture: https://github.com/open-craft/opencraft/tree/master/instance/tests/fixtures . Here at first sight you would probably want to mock both the responses the Django server sends to the client.

In any case please don't add any node/npm dependency, I beg you : )

omarkhan commented:
@itsjeyd this is ready for another look now. I have addressed your comments, as well as @smarnach's and @antoviaque's. @antoviaque wants me to contribute the email verification functionality to an existing lib and use that instead, so that's next on my list.

I have added js tests and selenium tests. The js tests will run when you run `make test_js`. For the selenium tests, run `make test_browser`.

omarkhan commented:
I can't reproduce a bug @itsjeyd reported. Can someone else see if they can reproduce it?

- Type some text into an input box
- Select the text and cut it. 'This field is required' should appear
- Paste the text back into the input box using the keyboard

For @itsjeyd the 'This field is required' message is still displayed. For me it disappears, as expected:
![registration](https://cloud.githubusercontent.com/assets/793379/14812539/ccff26ca-0bc7-11e6-8721-9f9c63ad27a3.gif)

itsjeyd commented on betatest/tests/js/betatest_app_spec.js:
```diff
@@ -0,0 +1,82 @@
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+(function(){
```
@omarkhan Nit: Missing space between `()` and `{`

antoviaque commented:
@omarkhan congratz on getting the Js tests done :) Now you still need to add an integration test, you know that, right? :)

@open-craft/core I meant to mention this in the meeting yesterday, but here is just as well - when you send a PR sharing open craft im, please consider adding an integration test. Sometimes it might not make sense to do one, but then please explain the rationale for it in the PR description. The coverage of integration tests is quite low at the moment, we should be careful to add them, especially for user facing features. Otherwise soon we'll be needing QA testers :)

omarkhan commented:
@antoviaque I did write [integration tests](https://github.com/open-craft/opencraft/blob/omar/beta-registration/betatest/tests/browser/browser_registration.py). Did you have something else in mind?

antoviaque commented on betatest/tests/browser/browser_registration.py:
```diff
@@ -0,0 +1,83 @@
+# Tests #######################################################################
+
+class BetaTestBrowserTestCase(BetaTestApplicationViewMixin,
+                              StaticLiveServerTestCase):
```
Is it the expected practice to be mixing a View mixin with the TestCase? These are two different types of objects?

omarkhan replied:
Oops it's actually a test mixin, I just didn't name it properly. Fixed: 8131b8a

antoviaque commented on betatest/tests/browser/browser_registration.py:
```diff
@@ -0,0 +1,83 @@
+
+        return self.client.page_source
+
+    def assert_registration_fails(self, form_data):
```
What is being run by the test runner? just the `assert_*()` function? Why not use `test_*()` like in other test cases?

omarkhan replied:
The test methods are defined in the mixin here: https://github.com/open-craft/opencraft/blob/omar/beta-registration/betatest/tests/test_views.py#L39

The code in this class fills in and submits the form using selenium instead of sending the form data directly using the django test client.

antoviaque replied:
I see - nice, thanks for the explanation. I'm good then : ) Maybe just a short comment so that it's more obvious? With the `..TestMixin` class that's already much more clear, but it can't hurt.

antoviaque commented on betatest/tests/browser/browser_registration.py:
```diff
@@ -0,0 +1,83 @@
+                element.send_keys(value)
+
+        submit = form.find_element_by_tag_name('button')
+        if submit.is_enabled():
```
When is this run?

It would be good to test the expected results more - currently I only see testing the activation status of a button, and that the next page loads. What page should load, What messages should be displayed in each step?

antoviaque commented:
@omarkhan My bad, I had missed them - that will teach me to reply on my phone : )  Cool, thanks. I was looking for an `integration` folder, but it makes sense to handle the browser tests separately from the server integration tests. I had a few questions/comments, which I posted inline in that code.

omarkhan commented:
@itsjeyd this is ready for a final review now.

itsjeyd commented on betatest/migrations/0001_initial.py:
```diff
@@ -0,0 +1,28 @@
+                ('instance_name', models.CharField(help_text='The name of your institution, company or project.\n\nExample: Hogwarts Online Learning', max_length=255)),
+                ('public_contact_email', models.EmailField(help_text='The email your instance of Open edX will be using to send emails, and where your users should send their support requests.\n\nThis needs to be a valid email.', max_length=254)),
+                ('project_description', models.TextField(verbose_name='your project', help_text='What are you going to use the instance for? What are your expectations?')),
+                ('subscribe_to_updates', models.BooleanField(help_text='I want OpenCraft to keep me updated about the progress of the beta test, and send me an email occasionally about it.', default=False)),
```
@omarkhan Can you update this to match the new wording? I.e., "... and occasionally send me an email about it." (I think the help texts for the other fields should still match but I didn't double-check.)

omarkhan replied:
Good catch @itsjeyd, I had forgotten to update the migration. Fixed

smarnach commented:
@omarkhan I cannor reproduce the issue with "This field is required" continuing to be displayed even when there is data in the input field again.  I tried on Firefox 46 and Chrome 49, and tried pasting by Keyboard short cut and context menu as well as Linux middle-click pasting.  All combinations worked fine, and I couldn't reproduce any issue.

omarkhan commented:
Thanks @smarnach. @itsjeyd OK to merge once I get a green build and squash the commits?

itsjeyd commented:
@omarkhan 

> OK to merge once I get a green build and squash the commits?

Yup, this looks great :+1:

antoviaque commented:
@omarkhan I still need to do a product review before the merge given that there is user-facing functionality - I've planned it for today though.

omarkhan commented:
@antoviaque OK. I have squashed the commits.

antoviaque commented:
@omarkhan Looks good! Kudos, this is nice both from the technical and user perspective : )

The comments I had while testing were:
* "Please use a stronger password." => do we state anywhere what the definition of a strong password is? It should be added both to the description on the right-hand column, and to the error message.
* With the default dev environment configuration, submitting the form crashes with https://i.imgur.com/c1F4XcD.png - and the only way to go around it is to setup proper SMTP credentials, and changing the from email to match what is accepted by this mailer. Could you add `EMAIL_BACKEND` to the settings, defaulting to `django.core.mail.backends.console.EmailBackend` and to `django.core.mail.backends.smtp.EmailBackend` on console.plebia.org?
* Currently, the approach is that the submit button is disabled until all the required fields are filled or checked. Which is good, except for the fact that it currently doesn't provide the user with any way to know what is missing, if he doesn't fill all the required fields. For example, only one of the two checkboxes at the end is mandatory - but if I don't check any of them, nothing tells me explicitly which field I am missing, and I can't to try to submit it to get an explicit error. I can't think of a good way to provide the user with that feedback at the right time (and only at the right time) if the submit button is disabled - showing it all the time would add a lot of red to the whole form until it is filled; we could also just add a "* Required" label next to all the fields, but the user would still have to look through all of these to search for the one he missed. The only simple and straightforward way that I see is to allow to click on the button when not all the requirements are met, and to show the error message on the missing fields then. That's what was (implicitly) intended in the requirements for "_error message at the top of the page, and one next to the issue explaining the issue and how to fix it_". But maybe there is a better approach?

omarkhan commented:
@antoviaque I have addressed your feedback, can you take another look?

antoviaque commented:
@omarkhan Thank you! I'll try, but I'm not sure I will have time for another review today, especially because that means reprovisioning my vagrant VM. Although you will be updating it to the current version once https://github.com/open-craft/opencraft/pull/6 is merged - ping me when it's done that would help. And if I don't get time to rereview before the meeting, feel free to merge to avoid the spillover - we would then take on any additional changes in a future task.

antoviaque commented:
@omarkhan I would have had time now to review, but this still needs a rebase. I'll try to reprovision a VM with your branch in the meantime, but you will need the rebase to merge on time before the meeting - note that it will require rebuilding your VM, so make sure to start it ahead of the meeting, to avoid the spillover.

antoviaque commented on betatest/forms.py:
```diff
@@ -0,0 +1,199 @@
+    password_confirmation = forms.CharField(
+        widget=forms.PasswordInput,
+        help_text=('Please use a strong password: avoid common patterns and '
+                   'make it long enough to be difficult to crack.'),
```
By definition, I meant giving the precise constraints - like the minimum number of characters, if there needs to be a combination of special characters, etc.

smarnach replied:
@antoviaque I specifically asked in the task description _not_ to have this kind of restrictions, and instead use a library that estimates password strength based on entropy, since restrictions that require special characters etc. are annoying.

@bradenmacdonald You might want to chime in here. :)

antoviaque replied:
@smarnach The problem is that the user needs to know what are the constraints - I hadn't noticed this requirement, or more precisely didn't realize that it meant that we can't have a clear definition for the user of the constraint here. Can we provide an example of what will be considered strong enough, in a reliable manner? Like "For example, a 15 letters word or series of words"? Would that be reliable, or will that depend on the letters?

omarkhan replied:
@antoviaque some background if you're interested: https://tech.dropbox.com/2012/04/zxcvbn-realistic-password-strength-estimation/

smarnach replied:
Fifteen times the letter `a` certainly wouldn't be strong enough.  I think the nicest feedback we could give would be to include a progress bar that indicates how strong the entered password is, with a mark at the required strength.  Otherwise, we can just give the general advice to add more lettters, or a bigger diversity of letters.

omarkhan replied:
@antoviaque it all depends. 'qwertyuiopasdfghjklzxcvbnm' is more than 15 letters but would fail validation, as it should. A series of words would work but again it depends on the words.

How about this: 'try to use at least 10 characters and avoid common patterns such as "password" or "qwerty".'

antoviaque replied:
We need something reliable to give to the user, so he doesn't get stuck and get accurate feedback when something is wrong - if he enters a 10 character password, and it's not enough, what should he do? Add more characters? Change the words he is using in it, because they are too common? He won't know.

If we can't provide complete instructions & clear error messages, unless we find a better idea, we will need to change the validation method. Since it's bigger than what you can do now, you can leave it aside - I'll create a task in the backlog about this.

bradenmacdonald replied:
I strongly prefer entropy-based password validation, using a tool like `zxcvbn`. What most websites do that use it is simply show a client-side "strength bar" while the user enters their password. Then, as they type every single character in their password, they can instantly see if the password is strong enough or not, and they can get an intuitive sense of what makes a strong password.

In such cases, it is typically not necessary to provide any further guidance to the user. People are used to making up passwords and generally have an idea of how to make a password stronger. I don't think it hurts to provide some specific examples though, such as something like "An example of a strong password would be thinking of a sentence with eight or more words and using the first letter of each word in the sentence, e.g. 'I love taking online courses with Open edX!' becomes 'IltocwOe!'"

omarkhan replied:
I agree with Braden. The only problem is that the python port of zxcvbn
does not always give the same results as the js reference implementation.
When Tim was testing this he found a password that passed client-side
validation but failed on the server. It would not be acceptable for a
client-side password strength meter to say that a password is very strong
only for the server to reject it when the form is submitted.

Possible solutions:

- Fix the python port of zxcvbn
- Use pyv8 or similar to call the js version of zxcvbn on the server

smarnach replied:
I also think the strength bar Braden described would be a good solution (but I'm bad at UX).  I have some doubts that we want pyv8 on the server, since I'd expect it to be more trouble than it's worth.  Fixing the Python port would be a great solution though, if it's feasible.

I also just saw the xkcd included in the link Omar gave above.  It's a very good summary why I don't like the usual numeral/uppercase/symbol rules: https://xkcd.com/936/

antoviaque replied:
I still don't see a solution here for the problem I was pointing out:

> We need something reliable to give to the user, so he doesn't get stuck and can get accurate feedback when something is wrong - if he enters a 10 character password, and it's not enough, what should he do? Add more characters? Change the words he is using in it, because they are too common? He won't know.

The progress bars are better than what there is now, but they don't answer this.

I agree that entropy based validation is better technically, and I would like to keep it. But users come first, so we would only keep it if we can avoid putting them in a situation where we tell them they are doing something wrong, without pointing out what, exactly. Especially in a critical step such as registration.

omarkhan replied:
The js implementation of zxcvbn includes nice user-readable messages
explaining how a password can be made stronger. These messages are absent
from the python port. We could port them over.

bradenmacdonald replied:
Alternately, you could only validate the password strength on the client side for now. If the user deliberately hacks around your JS validation to submit a weak password, it'll be their own fault.

The client form could submit the entropy score along with the password, so that the server can at least verify that the client-side checks seem to have run, and the entropy is reasonable.

That way, you get nice suggestion messages, and don't have to worry about inconsistencies between JS and python implementations.

antoviaque replied:
@omarkhan @bradenmacdonald Ah, cool - I've had a look at https://dl.dropboxusercontent.com/u/209/zxcvbn/test/index.html and the messages displayed there would definitely solve the problem on my side. Thanks for this. I let you update OC-1587 once you agree on the exact approach?

omarkhan replied:
I agree with Braden. I have updated OC-1587

antoviaque replied:
@omarkhan Perfect - thanks for the update!

antoviaque commented:
Kudos for the upstream PR to django-agular - quickly merged and released, awesome : )

antoviaque commented on requirements.txt:
```diff
@@ -9,7 +10,7 @@ debtcollector==0.8.0
 dj-static==0.0.6
 Django==1.8.5
-django-angular==0.7.15
+-e git+https://github.com/open-craft/django-angular.git@add307dbaa5df05f5839361ba899d88e83d658b2#egg=django-angular
```
You can use 0.8.2 from https://github.com/jrief/django-angular/pull/256 now, right?

omarkhan commented:
@antoviaque rebased. I don't see anything in https://github.com/open-craft/opencraft/pull/6 that would cause any problems with this. Once I get a green build I will merge.

antoviaque commented:
@omarkhan Alright - but https://github.com/open-craft/opencraft/pull/59#discussion_r61754286 still needs to be addressed. Can you have a look at my question there? If there is no time that can be put in a follow-up task, but if it's manageable by a string change it would be simpler to do it now.
